crates/bevy_ecs/src/schedule_v3/executor/multi_threaded.rs

```diff
@@ -0,0 +1,498 @@
+                        }
+
+                        self.rebuild_active_access();
+                    }
+                }
+
+                // SAFETY: all systems have completed
+                let world = unsafe { &mut *world.get() };
+                Self::apply_system_buffers(&mut self.unapplied_systems, schedule, world);
+
+                debug_assert!(self.ready_systems.is_clear());
+                debug_assert!(self.running_systems.is_clear());
+                debug_assert!(self.unapplied_systems.is_clear());
+                self.active_access.clear();
+                self.completed_sets.clear();
+                self.completed_systems.clear();
+            };
+
+            #[cfg(feature = "trace")]
+            let executor_span = info_span!("schedule_task");
+            #[cfg(feature = "trace")]
+            let executor = executor.instrument(executor_span);
+            scope.spawn(executor);
+        });
+    }
+}
+
+impl MultiThreadedExecutor {
+    pub fn new() -> Self {
+        let (sender, receiver) = async_channel::unbounded();
+        Self {
+            system_task_metadata: Vec::new(),
+            sender,
+            receiver,
+            dependencies_remaining: Vec::new(),
+            active_access: default(),
+            local_thread_running: false,
+            exclusive_running: false,
+            completed_sets: FixedBitSet::new(),
+            completed_systems: FixedBitSet::new(),
+            ready_systems: FixedBitSet::new(),
+            running_systems: FixedBitSet::new(),
+            unapplied_systems: FixedBitSet::new(),
+        }
+    }
+
+    fn spawn_system_tasks<'scope>(
+        &mut self,
+        ready: &FixedBitSet,
+        scope: &Scope<'_, 'scope, ()>,
+        schedule: &'scope SystemSchedule,
+        cell: &'scope SyncUnsafeCell<World>,
+    ) {
+        for system_index in ready.ones() {
+            // systems can be skipped within this loop
+            if self.completed_systems.contains(system_index) {
+                continue;
+            }
+
+            // SAFETY: no exclusive system running
+            let world = unsafe { &*cell.get() };
+            if !self.can_run(system_index, schedule, world) {
+                // NOTE: exclusive systems with ambiguities are susceptible to
+                // being significantly displaced here (compared to single-threaded order)
+                // if systems after them in topological order can run
+                // if that becomes an issue, `break;` if exclusive system
+                continue;
+            }
+            if !self.should_run(system_index, schedule, world) {
+                continue;
+            }
+
+            if !self.system_task_metadata[system_index].is_exclusive {
+                self.spawn_system_task(scope, system_index, schedule, world);
+            } else {
+                // SAFETY: no other system running
+                let world = unsafe { &mut *cell.get() };
```

Comment:
If the outer `world` were used again then yes it would be available (note it isn't used best I can tell)

Reply:
> Is this actually sound?

If you reach this line, there should be no other systems running, we immediately `break` after spawning the system, and won't enter this loop again until the exclusive system completes.

Looking at it again though, I must have lost the check for "no other systems running" at some point.

@BoxyUwU should probably take a look at this.

Reply:
I cant speak to anything about the scheduler and if that actually means there is no other references to world anywhere else because i have no idea how this works lmao, but as long as the `&World` isnt used after the `&mut World` then the `&mut World` wont be considered to alias the `&World`.  

Rust would definitely _let_ you use the `&World` afterwards, which would introduce UB to this code but that's what you get when using `UnsafeCell` :sweat_smile: The `// SAFETY` probably ought to talk about the previously created `&World` and how it is not used after this `&mut` is created.

Reply:
Should be resolved now.